ChibiOS upgrade to version 21.11.5

andyp1per opened the pull request:
Upgrade ChibiOS to the latest stable version with support for hal v9.1 and H5

Currently boots on a Pixhawk6X and TBS Lucid H7

Want to try this on a large variety of different MCUs
 - [x] L431 (Matek L431 Periph)
 - [x] G4xx (Matek G474 Periph) (Hitec-Airspeed?)
 - [x] H757 (CubeOrangePlus)
 - [x] H743 (Boots!)
 - [x] F427 (Cube Black)
 - [x] F1xx (iomcu)
 - [x] F3xx
 - [x] F405 / F407 (SpeedyBee F405 v4)

andyp1per committed:
AP_HAL_ChibiOS: fixes for ChibiOS kernel v9

andyp1per commented:
> Does this require updates to the OpenOCD and Segger thread debugging?

Don't know as yet.

andyp1per commented:
> I'm all in favor of updating ChibiOS, and I see you've updated the submodule.. but shouldn't there be a pull-request in ardupilot/ChibiOS for this too? Where's the changelog for v8?

There's a branch, no PR as yet but I can put it up.

tridge commented:
@andyp1per what branch is this based on? is it a rebase?

tridge commented:
@andyp1per Peter and I tested on a KakuteF7 quad and it seems that ESC telemetry messages come through less frequently on mavlink with this PR. They are not as frequent as I would normally expect (expect 4Hz) on master too, but with this PR I was seeing an update every 10s or so, so seems to have gotten worse.

tpwrules commented:
Definitely want to hold this til 4.8, and debugging needs testing and possible fixing.

andyp1per committed:
ChibiOS: kernel v9

andyp1per committed:
hwdef: correct I2C IRQs for STM32G4

andyp1per committed:
AP_HAL_ChibiOS: use correct PLLSAI values for F469

andyp1per committed:
AP_HAL_ChibiOS: support I2Cv4 single DMA channel

The ChibiOS I2Cv4 LLD (used on STM32G0/G4/C0/U0/U3/H5/L4+) uses a
single DMA channel per I2C peripheral, shared between TX and RX,
configured via STM32_I2C_I2Cx_DMA_CHANNEL. The previous code
emitted the separate STM32_I2C_I2Cx_RX/TX_DMA_STREAM defines used
by I2Cv2/I2Cv3, which no longer satisfy the I2Cv4 driver.

Detect I2Cv4 MCUs from CHIBIOS_PLATFORM_MK, allocate a single DMA
channel per I2C peripheral in dma_resolver.py (no _RX/_TX split),
and generate HAL_I2Cn_CONFIG entries that feed the single channel
into the shared DMA slot with SHARED_DMA_NONE as the partner.

Also replace the unconditional STM32_I2C_USE_I2Cn TRUE lines in
stm32g4_mcuconf.h with #ifndef-guarded defaults to FALSE, so
boards only enable the I2C peripherals their hwdef actually uses.
Without this, the I2Cv4 driver demanded DMA channels for every
I2C peripheral on every G4 board regardless of pin usage.

andyp1per committed:
AP_HAL_ChibiOS: correct I2C IRQs for STM32L4+

Rename STM32_I2C_I2Cn_IRQ_PRIORITY to STM32_IRQ_I2Cn_PRIORITY to
match the naming used by the kernel v9 I2Cv4 driver, mirroring
the same rename already applied to stm32g4_mcuconf.h.

andyp1per commented:
ok this is a rebase of all our changes on top of the ChibiOS stable 21.11.5 branch. Should all be in there now. I think we will probably have to branch the current ChibiOS repo as 21.11.3 and then force push this to master

peterbarker commented:
OK for flight testing for people that are game.

Andy wants to know if we're OK to just cherry-pick our things on top - looks fine to Peter.  This will require a forced-push of the master branch.  origin/stable-21.11.3 would continue to be used for older versions of ArduPilot.

peterbarker commented:
```
pbarker@crun:~/rc/ardupilot-claude(master)$ Tools/scripts/list_hwdef_mcus.py | grep '^## '
## CKS32F407xx  (2 boards)
## STM32F100xB  (2 boards)
## STM32F103xB  (4 boards)
## STM32F105xC  (1 board)
## STM32F303xC  (2 boards)
## STM32F405xx  (63 boards)
## STM32F407xx  (8 boards)
## STM32F412Rx  (13 boards)
## STM32F427xx  (17 boards)
## STM32F469xx  (1 board)
## STM32F732xx  (1 board)
## STM32F745xx  (10 boards)
## STM32F767xx  (10 boards)
## STM32F777xx  (3 boards)
## STM32G441xx  (1 board)
## STM32G474xx  (7 boards)
## STM32G491xx  (4 boards)
## STM32H723xx  (1 board)
## STM32H730xx  (1 board)
## STM32H743xx  (95 boards)
## STM32H750xx  (4 boards)
## STM32H753xx  (1 board)
## STM32H755xx  (1 board)
## STM32H757xx  (10 boards)
## STM32L431xx  (8 boards)
## STM32L476xx  (1 board)
## STM32L496xx  (1 board)
## STM32L4R5xx  (1 board)
pbarker@crun:~/rc/ardupilot-claude(master)$ 
```

peterbarker commented:
Flew this on a Bixler today without issue

andyp1per commented:
Thanks @peterbarker !

peterbarker commented:
```
## CKS32F407xx  (2 boards)
## STM32F100xB  (2 boards)
## STM32F103xB  (4 boards)
## STM32F105xC  (1 board)
## STM32F303xC  (2 boards)
## STM32F405xx  (63 boards)
## STM32F407xx  (8 boards)
## STM32F412Rx  (13 boards)
## STM32F427xx  (17 boards)
## STM32F469xx  (1 board)
## STM32F732xx  (1 board)
## STM32F745xx  (10 boards)
## STM32F767xx  (10 boards)
## STM32F777xx  (3 boards)
## STM32G441xx  (1 board)
## STM32G474xx  (7 boards)
## STM32G491xx  (4 boards)
## STM32H723xx  (1 board)
## STM32H730xx  (1 board)
## STM32H743xx  (95 boards)
## STM32H750xx  (4 boards)
## STM32H753xx  (1 board)
## STM32H755xx  (1 board)
## STM32H757xx  (10 boards)
## STM32L431xx  (8 boards)
## STM32L476xx  (1 board)
## STM32L496xx  (1 board)
## STM32L4R5xx  (1 board)
```

andyp1per commented:
@tridge I don't have an F3 to test - maybe you do?

tridge commented:
> @tridge I don't have an F3 to test - maybe you do?

I've tested a F303-GPS, works fine

peterbarker commented:
I've been flying this on my daily flyer Bixler for weeks now

peterbarker commented:
Oh, whoops.  Have we not merged the ChibiOS PR?!

peterbarker commented:
> Oh, whoops. Have we not merged the ChibiOS PR?!

Ah.  Of course, these are patches which are deliberately not on top of master, so I believe it's OK to merge this over the objections of CI.  We're applying ArduPilot's patches on top of a new base, so the test failing is fine.

We just want to update "master" in the repo to point to the new tip ASAP.

Am I thinking right?

andyp1per commented:
@peterbarker I have created the branch ArduPilot-4.7 to match current master in ChibiOS we then want to force push stable-21.11.5 over master to make it the latest, but I will let you or @tridge do that!
